### PR TITLE
Support non tuple return values from torch.nn.module

### PR DESCRIPTION
**Pull request by baijumeswani**
Support for following types is added:
- Any dictionary type (where values are ```torch.Tensor``` type)
- Named tuples
- HuggingFace ModelOutput types (where values are only ```torch.Tensor``` type. Some of the outputs contain tuples of tensors which are not yet supported.)

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule.py
Comment (mrry) [inline orttraining/orttraining/python/training/ortmodule.py:455]: Could `self._sample_user_output` contain large tensors here? Unless I'm mistaken I *think* all we need to keep is `type(sample_user_output)`.

**Review by mrry — CHANGES_REQUESTED** [review]
Nice fix! I have just one suggestion that might make it a little more efficient, but the overall shape looks good to me.

File: orttraining/orttraining/python/training/ortmodule.py
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule.py:455]: You're right. I had initially started off with just saving the type. Later, I was experimenting with having more complex return types (such as dictionary of dictionaries ...), which was when I changed this to save the entire sample output.
But I think since this PR only addresses the simple types, saving just the type is enough. Will update the PR with the change now while I think of ways to save types of more complex outputs.

File: orttraining/orttraining/python/training/ortmodule.py
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:458]: I would suggest a name that resembles more to the original model, such as `self._original_module_output_type`  or maybe `self._original_module_user_output_type` which as associated with `self._original_module`

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:399]: How about some tests that handle unsupported types?

**Comment by ytaous** [comment]
Hi, I was testing gtp-2 example using dev branch, and got the error below.

File "/bert_ort/ettao/ettao-py36/lib/python3.6/site-packages/onnxruntime/training/ortmodule.py", line 268, in forward
    self._get_forward_graph_and_init_gradient_graph_builder(*inputs, **kwargs)
  File "/bert_ort/ettao/ettao-py36/lib/python3.6/site-packages/onnxruntime/training/ortmodule.py", line 158, in _get_forward_graph_and_init_gradient_graph_builder
    self._onnx_training = self._get_forward_graph(input_names, dynamic_axes, *inputs, **kwargs)
...
...
  File "/bert_ort/ettao/ort-example/onnxruntime-training-examples/huggingface-gpt2/transformers/src/transformers/modeling_gpt2.py", line 436, in forward
    raise ValueError("You have to specify either input_ids or inputs_embeds")


So I tried [https://github.com/microsoft/onnxruntime/pull/6539](https://github.com/microsoft/onnxruntime/pull/6539), but got different error:

  File "/bert_ort/ettao/ettao-py36/lib/python3.6/site-packages/onnxruntime/training/ortmodule.py", line 152, in _parse_outputs_for_onnx_export
    tmp_output_names, tmp_output_dynamic_axes = _create_output_dim_names(out, idx, True)
  File "/bert_ort/ettao/ettao-py36/lib/python3.6/site-packages/onnxruntime/training/ortmodule.py", line 120, in _create_output_dim_names
    raise TypeError('ORTModule does not support the following model output type {} within a Sequence'.format(type(sample_outputs)))
TypeError: ORTModule does not support the following model output type <class 'tuple'> within a Sequence


I heard the issue should go away after this PR is merged. But I am curious if I just clone this private branch to try and see what happen. 
Turns out I am getting the the first error above. So I wonder if I should wait for both PR merged to test again?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py
Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:399]: I'll add a unit test for that.

**Comment by baijumeswani** [comment]
@ytaous I don't believe this PR would help you resolve the error you're seeing. From the error message you posted, it appears that the the output of the model is a sequence (list, tuple ...) of non tensor values.
And this hierarchical/nested structure is not supported at the moment. I will reach out to you to discuss what we can possibly do to mitigate or work around the problem you're seeing.

**Review by thiagocrepaldi — APPROVED** [review]

**Pull request merged**
